spring-cloud-dataflow-admin/src/test/java/org/springframework/cloud/dataflow/admin/controller/StreamControllerTests.java

```diff
@@ -0,0 +1,201 @@
+						.andExpect(status().isCreated());
+		assertEquals(1, repository.count());
+		StreamDefinition myStream = repository.findOne("myStream");
+		assertEquals("time | log", myStream.getDslText());
+		assertEquals("myStream", myStream.getName());
+		assertEquals(2, myStream.getModuleDefinitions().size());
+		ModuleDefinition timeDefinition = myStream.getModuleDefinitions().get(0);
+		ModuleDefinition logDefinition = myStream.getModuleDefinitions().get(1);
+		assertEquals(1, timeDefinition.getParameters().size());
+		assertEquals("myStream.0", timeDefinition.getParameters().get(BindingProperties.OUTPUT_BINDING_KEY));
+		assertEquals(1, logDefinition.getParameters().size());
+		assertEquals("myStream.0", logDefinition.getParameters().get(BindingProperties.INPUT_BINDING_KEY));
+	}
+
+	@Test
+	public void testSaveDuplicate() throws Exception {
+		repository.save(new StreamDefinition("myStream", "time | log"));
+		assertEquals(1, repository.count());
+		mockMvc.perform(
+				post("/streams/definitions/").param("name", "myStream").param("definition", "time | log")
+						.accept(MediaType.APPLICATION_JSON)).andDo(print())
+						.andExpect(status().is5xxServerError());
+		assertEquals(1, repository.count());
+	}
+
+	@Test
+	public void testSaveWithParameters() throws Exception {
+		assertEquals(0, repository.count());
+		String definition = "time --fixedDelay=500 --timeUnit=milliseconds | log";
+		mockMvc.perform(
+				post("/streams/definitions/").param("name", "myStream")
+						.param("definition", definition)
+						.accept(MediaType.APPLICATION_JSON)).andDo(print())
+				.andExpect(status().isCreated());
+		assertEquals(1, repository.count());
+		StreamDefinition myStream = repository.findOne("myStream");
+		ModuleDefinition timeDefinition = myStream.getModuleDefinitions().get(0);
+		ModuleDefinition logDefinition = myStream.getModuleDefinitions().get(1);
+		assertEquals("time", timeDefinition.getName());
+		assertEquals("log", logDefinition.getName());
+		assertEquals("500", timeDefinition.getParameters().get("fixedDelay"));
+		assertEquals("milliseconds", timeDefinition.getParameters().get("timeUnit"));
+		assertEquals(definition, myStream.getDslText());
+		assertEquals("myStream", myStream.getName());
+	}
+
+	@Test
+	public void testDestroyStream() throws Exception {
+		repository.save(new StreamDefinition("myStream", "time | log"));
+		assertEquals(1, repository.count());
+		mockMvc.perform(
+				delete("/streams/definitions/myStream").accept(MediaType.APPLICATION_JSON)).andDo(print())
+				.andExpect(status().isOk());
+		assertEquals(0, repository.count());
+	}
+
+	@Test
+	public void testDestroyStreamNotFound() throws Exception {
+		mockMvc.perform(
+				delete("/streams/definitions/myStream").accept(MediaType.APPLICATION_JSON)).andDo(print())
+				.andExpect(status().is5xxServerError());
+		assertEquals(0, repository.count());
+	}
+
+	@Test
+	@SuppressWarnings({"unchecked", "rawtypes"})
+	public void testDeploy() throws Exception {
```

Comment:
`StreamController` works with `ModuleDeployer`, not `ModuleLauncher`. It seems to me that this test is assuming too much knowledge about the local module deployer. Instead of mocking `ModuleLauncher` should it mock `ModuleDeployer` and assert that the correct `ModuleDeploymentRequest`s are being created?